### PR TITLE
Fix UINT8 overflow for >255 segmentation masks

### PR DESCRIPTION
**Pull request by rajeshtims**
Like the title suggests, when the number of objects in an image was more than 255, it caused uint8 overflow. [Line 219] To fix this, dtype of the a mask appended to the list *masks* was changed to uint32. 
This alone was not enough fix as the datatype of *masks* list could be assigned to uint8 on line 213. I changed the conditional datatype to int32. 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to data type handling and file validation process.

### 📊 Key Changes
- Updated the `polygons2masks_overlap` function to ensure masks have consistent data types.
- Improved file extension checking logic in the validator by using `str.endswith(".yaml")`.

### 🎯 Purpose & Impact
- 🛠️ **Improved Consistency**: Ensures mask data types are consistent, reducing potential errors in mask operations.
- 🧠 **Better Validation**: Simplifies YAML file checks, enhancing code readability and maintainability.
- ⚠️ **User Alert**: Provides a clearer warning for users about the validation of untrained models, preventing potential confusion.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rajeshtims, thank you for submitting your PR to the `ultralytics/ultralytics` repository! 🚀 To help us integrate your valuable contributions smoothly, please consider the following checklist:

- ✅ **Purpose Explanation**: Ensure your PR description clearly outlines the purpose of your fix related to uint8 overflow, and links to any relevant [issues](https://github.com/ultralytics/ultralytics/issues) if applicable. Make sure your commit messages follow our project conventions.
- ✅ **Synchronization**: Verify your PR is up-to-date with the `main` branch. If necessary, update it using the 'Update branch' feature or by running `git pull` and `git merge main` locally.
- ✅ **CI Verification**: Confirm that all required [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are successful. Address any failed checks promptly.
- ✅ **Documentation Update Required**: Please update any necessary [documentation](https://docs.ultralytics.com) to reflect the changes made.
- ✅ **Testing**: Add or update relevant tests to cover the changes introduced by your PR. Make sure all tests pass successfully.
- ✅ **Sign the CLA**: If this is your first contribution, don't forget to sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/). Acknowledge this by stating, "I have read the CLA Document and I sign the CLA" in the comments.
- ✅ **Streamline Modifications**: Aim to keep changes minimal and essential. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For any additional details, please review our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Feel free to leave a comment if you need further assistance—our engineering team will be in touch soon. Thank you for contributing to Ultralytics! 🌟🥳

**Comment by rajeshtims** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@rajeshtims thanks for the PR! We used lower quantization to avoid the performance hit of 32 bits. Could we simply update the lower type cast to the current masks type instead of fixing all to 32?

File: ultralytics/data/utils.py [hunk 1]
```diff
@@ -210,13 +210,13 @@ def polygons2masks_overlap(imgsz, segments, downsample_ratio=1):
...
         (imgsz[0] // downsample_ratio, imgsz[1] // downsample_ratio),
-        dtype=np.int32 if len(segments) > 255 else np.uint8,
+        dtype=np.int32,
     )
     areas = []
     ms = []
     for si in range(len(segments)):
         mask = polygon2mask(imgsz, [segments[si].reshape(-1)], downsample_ratio=downsample_ratio, color=1)
-        ms.append(mask)
+        ms.append(mask.astype(np.uint32))
```
Comment (glenn-jocher) [inline ultralytics/data/utils.py:219]: Maybe change L219 instead?
```
ms.append(mask.astype(masks.dtype))
```

**Comment by glenn-jocher** [comment]
Ok trying my little experiment

**Comment by glenn-jocher** [comment]
@rajeshtims looks like it's working. I'll merge this version, and pleae open a new PR if this doesn't resolve your issue. Should be live in `ultralytics 8.3.5` shortly.

**Pull request merged**